### PR TITLE
4032 - Update collaborators view

### PR DESCRIPTION
**Pull request by sorja**
This PR concludes Collaborator view refactor

- [x] Test all roles
Screenshot 2024-11-22 at 13 52 15

### CODE REVIEW

File: src/server/repository/public/user/getMany.ts [hunk 1]
```diff
@@ -33,7 +33,11 @@ const _getOrderClause = (
   orderBy: string | undefined = 'full_name',
   orderByDirection: TablePaginatedOrderByDirection | undefined = TablePaginatedOrderByDirection.asc
 ): string => {
-  return `order by ${orderBy} ${orderByDirection}`
+  // Sort users by whether they have roles or invitations, then by the requested sort field
+  return `order by
+    case when (select jsonb_array_length(coalesce(jsonb_agg(role) filter (where role is not null), '[]'))) > 0 then 0 else 1 end,
+    case when (select jsonb_array_length(coalesce(jsonb_agg(invitation) filter (where invitation is not null), '[]'))) > 0 then 0 else 1 end,
```
Comment (sorja) [inline src/server/repository/public/user/getMany.ts]: There is probably a better way, but I wanted to use the same clause we use in the view itself to create the roles array

File: src/client/pages/CountryHome/Collaborators/UserList/Users/UserCard/Actions/Actions.tsx [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import Edit from './Edit'
+import Message from './Message'
+import Remove from './Remove'
+
+type ActionType = {
+  Component: React.FC<Props>
+  name: string
+}
+
+const useActions = (props: Props) => {
```
Comment (minotogna) [inline src/client/pages/CountryHome/Collaborators/UserList/Users/UserCard/Actions/Actions.tsx]: Better to extract separate file and add return type

File: src/client/pages/CountryHome/Collaborators/UserList/Users/UserCard/Actions/hooks/useRefetchUsers.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import { useAppDispatch } from 'client/store'
+import { TablePaginatedActions } from 'client/store/ui/tablePaginated'
+import { useCountryRouteParams } from 'client/hooks/useRouteParams'
+
+export const useRefetchUsers = () => {
+  const { assessmentName, cycleName, countryIso } = useCountryRouteParams<CountryIso>()
+
+  const dispatch = useAppDispatch()
+
+  const getDataProps = useMemo(() => {
```
Comment (minotogna) [inline src/client/pages/CountryHome/Collaborators/UserList/Users/UserCard/Actions/hooks/useRefetchUsers.ts]: Add type , always better

File: src/client/pages/CountryHome/Collaborators/UserList/Users/UserCard/Actions/hooks/useRefetchUsers.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import { useCallback, useMemo } from 'react'
+
+import { ApiEndPoint } from 'meta/api/endpoint'
+import { CountryIso } from 'meta/area'
+
+import { useAppDispatch } from 'client/store'
+import { TablePaginatedActions } from 'client/store/ui/tablePaginated'
+import { useCountryRouteParams } from 'client/hooks/useRouteParams'
+
+export const useRefetchUsers = () => {
```
Comment (minotogna) [inline src/client/pages/CountryHome/Collaborators/UserList/Users/UserCard/Actions/hooks/useRefetchUsers.ts]: Return type

File: src/client/pages/CountryHome/Collaborators/UserList/Users/UserCard/Avatar/Avatar.tsx [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import './Avatar.scss'
+import React from 'react'
+
+import { CountryUserSummary, Users } from 'meta/user'
+
+type Props = {
+  user: CountryUserSummary
+}
+
+const Avatar: React.FC<Props> = (props: Props) => {
```
Comment (minotogna) [inline src/client/pages/CountryHome/Collaborators/UserList/Users/UserCard/Avatar/Avatar.tsx]: Don't we have already a user avatar common component ? If not , we should ?

File: src/client/pages/CountryHome/Collaborators/UserList/Users/UserCard/Info/Info.scss [hunk 6]
```diff
@@ -0,0 +1,48 @@
+@import 'src/client/style/partials';
+
```
Comment (minotogna) [inline src/client/pages/CountryHome/Collaborators/UserList/Users/UserCard/Info/Info.scss:2]: These css class names are too generic. The context is missing .
home-users ?

File: src/client/pages/CountryHome/Collaborators/UserList/Users/UserCard/UserCard.scss [hunk 7]
```diff
@@ -0,0 +1,19 @@
+@import 'src/client/style/partials';
+
+.user-card {
```
Comment (minotogna) [inline src/client/pages/CountryHome/Collaborators/UserList/Users/UserCard/UserCard.scss]: Also here

File: src/client/pages/CountryHome/Collaborators/UserList/Users/Users.scss [hunk 8]
```diff
@@ -1,22 +1,11 @@
 @import 'src/client/style/partials';
 
 .users {
```
Comment (minotogna) [inline src/client/pages/CountryHome/Collaborators/UserList/Users/Users.scss]: And this

File: src/client/pages/CountryHome/Collaborators/UserList/Users/Users.scss [hunk 9]
```diff
@@ -1,22 +1,11 @@
...
-    color: $text-mute;
-    font-style: italic;
-  }
+  .table-paginated-datagrid {
+    grid-row-gap: $spacing-xxs;
 
-  .expired {
-    color: lighten($ui-destructive, 40%);
+    .data-grid-column.withBorder {
+      //border: unset;
```
Comment (minotogna) [inline src/client/pages/CountryHome/Collaborators/UserList/Users/Users.scss]: Should we remove this ?

File: src/server/repository/public/user/getMany.ts (hunk 1 above)
Comment (minotogna) [reply src/server/repository/public/user/getMany.ts]: should we sort this client side ? and only in country home users list.
But it should be generic.
maybe we could pass an optional sortBy function to TablePaginated ? 

What do you think?

Comment (sorja) [reply src/server/repository/public/user/getMany.ts]: Sure, removed backend custom sorting and added `sortFn: SortFn<Datum>` prop for TablePaginated. Seems to work nicely, let me know

Sort prop: https://github.com/openforis/fra-platform/pull/4139/commits/38b9cdb566887fb05fb073c5c45e1e7f9c077a11

Remove backend sorting and add sortFn for users: https://github.com/openforis/fra-platform/pull/4139/commits/afbfaa28b6f4d637c6bfcad1ccadddec644abd08

**Review by minotogna — APPROVED** [review]
👍

File: src/server/repository/public/user/getMany.ts (hunk 1 above)
Comment (minotogna) [reply src/server/repository/public/user/getMany.ts]: nice 👍

**Pull request merged**